File: aoc-dec03/dec03.jl
using SparseArrays

function preparemap(thismap)
    splitmap = split(thismap)
    J = length(splitmap)
    I = length(first(splitmap))
    sparse(reshape(vcat([[c=='#' for c in l] for l in split(thismap)]...),(I,J)))
end

function is_tree(preparedmap, pos::Pair)
    is_tree(preparedmap, pos.first, pos.second)
end

function is_tree(preparedmap, x, y)
    X,Y = size(preparedmap)
    preparedmap[mod1(x,X), y]
end

function next_pos(pos, delta)
    Pair(pos.first + delta.first, pos.second + delta.second)
end

function traverse(thismap, start, next_pos)
    X,Y = size(thismap)
    num_trees = is_tree(thismap, start) ? 1 : 0
    if start.second < Y
        num_trees + traverse(thismap, next_pos(start), next_pos)
    else
        num_trees
    end
end

# Part I
function dec03_part1(preparedmap)
    traverse(preparedmap, (1=>1), pos->next_pos(pos,(3=>1)))
end

# Part II
function dec03_part2(preparedmap)
    slopes = ((1=>1), (3=>1), (5=>1), (7=>1), (1=>2))
    prod((traverse(p, (1=>1), pos->next_pos(pos,s)) for s in slopes))
end

function dec03_both(preparedmap, slopes)
    prod((traverse(p, (1=>1), pos->next_pos(pos,s)) for s in slopes))
end

inputdata = read(joinpath(@__DIR__,"input.txt"),String);
p = preparemap(inputdata);
slopes = ((1=>1), (3=>1), (5=>1), (7=>1), (1=>2));

using Test
@test dec03_part1(p) == 299
@test dec03_part2(p) == 3621285278
@test dec03_both(p, [slopes[2]]) == 299
@test dec03_both(p, slopes) == 3621285278

File: aoc-dec03/dec03.clj
(require '[clojure.string :as str])

(defn is-tree [thismap x y] ;1-indexed positions x and y
    (def splitmap (str/split-lines thismap))
    (def Y (count splitmap))
    (def X (count (get splitmap 0)))
    (if
    (= (get (get splitmap (dec y)) (mod (dec x) X)) (get "#" 0))
    1
    0
    )
)

(defn count-path [thismap [x y]]
    (def Y (count (str/split-lines thismap)))
    (def has-tree (is-tree thismap x y))
    (println x y has-tree)
    (if (< y Y)
        (+ has-tree (count-path thismap (next-pos [x y])))
        (is-tree thismap x y);has-tree
    )
)

(defn next-pos
  [[x y] [dx dy]]
  [(+ x dx) (+ y dy)]
  )

(defn count-path [thismap [x y] next_pos]
    (def Y (count (str/split-lines thismap)))
    (def has-tree (is-tree thismap x y))
    ; (println x y has-tree)
    (if (< y Y)
        (+ has-tree (count-path thismap (next_pos [x y]) next_pos))
        (is-tree thismap x y);has-tree
    )
)

(defn dec03-1 [thismap]
  (count-path thismap [1 1] (fn [[x y]] (next-pos [x y] [3 1])))
  )


(defn dec03-2 [thismap]
    (def slopes [[1 1] [3 1] [5 1] [7 1] [1 2]])
    (reduce * (for [s slopes] (count-path thismap [1 1] (fn [[x y]] (next-pos [x y] s)))))
)

(def inputdata (slurp "input.txt"))

